### PR TITLE
[MAINTENANCE] Add tests around expectations

### PR DESCRIPTION
**Pull request by tyler-hoffman**
This PR is meant in part to highlight my strategy for the expectation tests. Some things to call out (feel free to push back):
* 1:1 relationship between expectation and test file
* Run as few expectations against the full set of datasources <- I suspect this gives us reasonable confidence, but maybe we should just run them all against all backends?
* Less combinatorial explosion than what we had (e.g. the old json file tested the case when we had duplicates, the case when there were nulls, and the case when there were duplicates and nulls; I see no benefit to the third)
* Same idea as ^ for testing details, e.g. only one test to cover the COMPLETE results
* I'm not testing all of types that are supported here. My thinking as of now is that CORE-599 can test the details around what backends support which types, e.g. I'm not testing that we support DATE and DATETIME or INT and FLOAT.

- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, visit our [community resources](https://docs.greatexpectations.io/docs/core/introduction/community_resources#contribute-code-or-documentation).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: tests/integration/data_sources_and_expectations/test_canonical_expectations.py [hunk 1]
```diff
@@ -33,6 +33,8 @@
     SqliteDatasourceTestConfig(),
 ]
 
+JUST_PANDAS_DATA_SOURCES: Sequence[DataSourceTestConfig] = [PandasDataFrameDatasourceTestConfig()]
```
Comment (tyler-hoffman) [inline tests/integration/data_sources_and_expectations/test_canonical_expectations.py:63]: I think we'll eventually phase out this file and move this and the other constant list out, but for now, this seemed like the easiest place to put this.

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+ONES_AND_TWOS = pd.DataFrame({COL_NAME: [1, 2, 2, 2]})
+
+
+@parameterize_batch_for_data_sources(data_source_configs=ALL_DATA_SOURCES, data=ONES_AND_TWOS)
+def test_success_complete_results(batch_for_datasource: Batch) -> None:
+    expectation = gxe.ExpectColumnDistinctValuesToBeInSet(column=COL_NAME, value_set=[1, 2])
+    result = batch_for_datasource.validate(expectation, result_format=ResultFormat.COMPLETE)
+    assert result.success
+    assert result.to_json_dict()["result"] == {
```
Comment (tyler-hoffman) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:26]: This is ugly, but we need to call `to_json_dict()` to deal with some dataframe comparison issues.

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        "details": {
+            "value_counts": [
+                {"value": 1, "count": 1},
```
Comment (tyler-hoffman) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:29]: Why didn't we design this so that `value_counts` was just a dictionary of value -> count, rather than a list of these?

File: great_expectations/expectations/core/expect_column_kl_divergence_to_be_less_than.py [hunk 4]
```diff
@@ -327,9 +327,13 @@ class ExpectColumnKLDivergenceToBeLessThan(ColumnAggregateExpectation):
     tail_weight_holdout: Union[float, None] = pydantic.Field(
         default=0, ge=0, le=1, description=TAIL_WEIGHT_HOLDOUT_DESCRIPTION
     )
-    bucketize_data: bool = pydantic.Field(True, description=BUCKETIZE_DATA_DESCRIPTION)
-    min_value: Optional[Comparable] = pydantic.Field(None, description=MIN_VALUE_DESCRIPTION)
-    max_value: Optional[Comparable] = pydantic.Field(None, description=MAX_VALUE_DESCRIPTION)
+    bucketize_data: bool = pydantic.Field(default=True, description=BUCKETIZE_DATA_DESCRIPTION)
+    min_value: Optional[Comparable] = pydantic.Field(
+        default=None, description=MIN_VALUE_DESCRIPTION
```
Comment (joshua-stauffer) [inline great_expectations/expectations/core/expect_column_kl_divergence_to_be_less_than.py:332]: i'm curious, why all the updates to the kwarg `default=None`? i appreciate it for readability, but wondering if there's a functional reason.

Comment (tyler-hoffman) [reply great_expectations/expectations/core/expect_column_kl_divergence_to_be_less_than.py:332]: Responded in slack too with the longer answer, but it's to make mypy understand

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    assert result.success
+
+
+@parameterize_batch_for_data_sources(
+    data_source_configs=JUST_PANDAS_DATA_SOURCES, data=pd.DataFrame({COL_NAME: [1, 2, None]})
+)
+def test_ignores_nulls(batch_for_datasource: Batch) -> None:
+    expectation = gxe.ExpectColumnDistinctValuesToBeInSet(column=COL_NAME, value_set=[1, 2])
+    result = batch_for_datasource.validate(expectation)
+    assert result.success
```
Comment (joshua-stauffer) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:68]: very weird that this is our goal behavior

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    expectation = gxe.ExpectColumnDistinctValuesToBeInSet(column=COL_NAME, value_set=[])
+    result = batch_for_datasource.validate(expectation)
+    assert not result.success
+
+
+@parameterize_batch_for_data_sources(
+    data_source_configs=JUST_PANDAS_DATA_SOURCES, data=ONES_AND_TWOS
+)
+def test_value_set_is_none(batch_for_datasource: Batch) -> None:
+    # why do we even allow this?!?
```
Comment (joshua-stauffer) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:95]: i'm here for the commentary

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py (hunk 5 above)
Comment (tyler-hoffman) [reply tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:68]: Yeah, we're pretty consistent about omitting nulls from the comparisons.

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+def test_ignores_nulls(batch_for_datasource: Batch) -> None:
+    expectation = gxe.ExpectColumnDistinctValuesToBeInSet(column=COL_NAME, value_set=[1, 2])
+    result = batch_for_datasource.validate(expectation)
+    assert result.success
+
+
+@parameterize_batch_for_data_sources(
+    data_source_configs=JUST_PANDAS_DATA_SOURCES, data=ONES_AND_TWOS
+)
+def test_data_is_subset(batch_for_datasource: Batch) -> None:
```
Comment (joshua-stauffer) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:74]: why does this only apply to pandas?

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    # why do we even allow this?!?
+    expectation = gxe.ExpectColumnDistinctValuesToBeInSet(column=COL_NAME, value_set=None)
+    result = batch_for_datasource.validate(expectation)
+    assert result.success
+
+
+@parameterize_batch_for_data_sources(
+    data_source_configs=JUST_PANDAS_DATA_SOURCES, data=ONES_AND_TWOS
+)
+def test_failure(batch_for_datasource: Batch) -> None:
```
Comment (joshua-stauffer) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:104]: also curious here - why not test failure against all data sources?

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py (hunk 7 above)
Comment (tyler-hoffman) [reply tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:74]: I've been mostly covering the golden path with all data sources, as that should cover the broad metrics, and then covering more edge cases with just pandas since we don't have performance concerns there. Happy to revisit that if you think otherwise.

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+import pytest
+
+import great_expectations.expectations as gxe
+from great_expectations.core.result_format import ResultFormat
+from great_expectations.datasource.fluent.interfaces import Batch
+from tests.integration.conftest import parameterize_batch_for_data_sources
+from tests.integration.data_sources_and_expectations.test_canonical_expectations import (
+    ALL_DATA_SOURCES,
+    DATA_SOURCES_THAT_SUPPORT_DATE_COMPARISONS,
+    JUST_PANDAS_DATA_SOURCES,
```
Comment (cdkini) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:13]: Should we put these groupings in some kind of shared file or conftest?

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_contain_set.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    assert result.success
+
+
+@parameterize_batch_for_data_sources(
+    data_source_configs=JUST_PANDAS_DATA_SOURCES, data=ONES_AND_TWOS
+)
+def test_failure(batch_for_datasource: Batch) -> None:
+    expectation = gxe.ExpectColumnDistinctValuesToContainSet(column=COL_NAME, value_set=[1, 2, 3])
+    result = batch_for_datasource.validate(expectation)
+    assert not result.success
```
Comment (cdkini) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_contain_set.py:83]: On failure cases should we check anything else?

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_most_common_value_to_be_in_set.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    "expectation",
+    [
+        pytest.param(
+            gxe.ExpectColumnMostCommonValueToBeInSet(
+                column=COL_WITH_TIE, value_set=[100, 101], ties_okay=True
+            ),
+            id="no_matches_to_value_set",
+        ),
+        pytest.param(
+            # This feels like it conflicts with the 2 positive cases above
```
Comment (joshua-stauffer) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_most_common_value_to_be_in_set.py:83]: agree, what's the process for following up with stuff that's inconsistent, just weird, or maybe wrong? have you made tickets for any of these?

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_pair_values_a_to_be_greater_than_b.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+NULLS_B = "nulls_b"
+NULLS_C = "nulls_c"
+
+DATA = pd.DataFrame(
+    {
+        NUMBERS_A: [5, 8, 11, 13, 14],
+        NUMBERS_A_EQUAL: [5, 8, 11, 13, 14],
+        NUMBERS_B: [1, 5, 9, 12, 12],
+        DATES_A: [
+            datetime(2024, 1, 1).date(),  # noqa: DTZ001
```
Comment (joshua-stauffer) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_pair_values_a_to_be_greater_than_b.py:34]: i see a lot of `DTZ001`, why not just pass in `tz=timezone.utc()` instead?

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_proportion_of_unique_values_to_be_between.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    MSSQLDatasourceTestConfig(column_types=COLUMN_TYPES),
+    MySQLDatasourceTestConfig(column_types=COLUMN_TYPES),
+    PandasDataFrameDatasourceTestConfig(),
+    PandasFilesystemCsvDatasourceTestConfig(),
+    PostgreSQLDatasourceTestConfig(column_types=COLUMN_TYPES),
+    SnowflakeDatasourceTestConfig(column_types=COLUMN_TYPES),
+    SparkFilesystemCsvDatasourceTestConfig(),
+    SqliteDatasourceTestConfig(column_types=COLUMN_TYPES),
+]
+
```
Comment (joshua-stauffer) [inline tests/integration/data_sources_and_expectations/expectations/test_expect_column_proportion_of_unique_values_to_be_between.py:65]: why do you redefine this here?

Comment (joshua-stauffer) [reply tests/integration/data_sources_and_expectations/expectations/test_expect_column_proportion_of_unique_values_to_be_between.py:65]: required for column types? if so, maybe add a small comment explaining why the pattern changes for this case, since its not necessarily obvious.

Comment (tyler-hoffman) [reply tests/integration/data_sources_and_expectations/expectations/test_expect_column_proportion_of_unique_values_to_be_between.py:65]: Yeah, required for col types. I've since found what I think is a cleaner solution to this that I can migrate to. If that doesn't work, I'll comment. I'll leave this convo unresolved so I remember to get to it.

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py (hunk 7 above)
Comment (joshua-stauffer) [reply tests/integration/data_sources_and_expectations/expectations/test_expect_column_distinct_values_to_be_in_set.py:74]: yeah that makes sense. I was wondering if we should rename/duplicate it to `THE_REALLY_FAST_ONES` or something and include Spark. But i guess the problem is that if an expectation doesn't support spark, it won't be clear from the variable name that we can't use it.

**Review by joshua-stauffer — APPROVED** [review]
pattern looks great ✅ left some comments just for understanding, but i think this is good to ship as is. thanks for doing this!

File: tests/integration/data_sources_and_expectations/expectations/test_expect_column_proportion_of_unique_values_to_be_between.py (hunk 13 above)
Comment (joshua-stauffer) [reply tests/integration/data_sources_and_expectations/expectations/test_expect_column_proportion_of_unique_values_to_be_between.py:65]: nice, thanks 🙇
